components/raftstore-v2/src/operation/command/write/raw_write.rs

```diff
@@ -0,0 +1,484 @@
+            }
+        } else {
+            return false;
+        };
+        true
+    }
+
+    #[inline]
+    pub fn amend(&mut self, req: RaftCmdRequest) -> Result<(), RaftCmdRequest> {
+        if Self::allow_request(&req) && req.header == self.header {
+            let last_length = self.buf.len();
+            for r in req.get_requests() {
+                encode(r, &mut self.buf);
+            }
+            // The default size limit is 8 * 0.4 = 3.2MiB.
+            if self.buf.len() < self.size_limit {
+                Ok(())
+            } else {
+                self.buf.truncate(last_length);
+                Err(req)
+            }
+        } else {
+            Err(req)
+        }
+    }
+
+    #[inline]
+    pub fn data_size(&self) -> usize {
+        self.buf.len()
+    }
+
+    #[inline]
+    pub fn encode(self) -> (Vec<u8>, Vec<CmdResChannel>) {
+        (self.buf, self.channels)
+    }
+
+    #[inline]
+    pub fn add_response_channel(&mut self, ch: CmdResChannel) {
+        self.channels.push(ch);
+    }
+}
+
+#[derive(Debug)]
+pub struct Put<'a> {
+    pub cf: &'a str,
+    pub key: &'a [u8],
+    pub value: &'a [u8],
+}
+
+#[derive(Debug)]
+pub struct Delete<'a> {
+    pub cf: &'a str,
+    pub key: &'a [u8],
+}
+
+#[derive(Debug)]
+pub struct DeleteRange<'a> {
+    pub cf: &'a str,
+    pub start_key: &'a [u8],
+    pub end_key: &'a [u8],
+    pub notify_only: bool,
+}
+
+#[derive(Debug)]
+pub enum RawWrite<'a> {
+    Put(Put<'a>),
+    Delete(Delete<'a>),
+    DeleteRange(DeleteRange<'a>),
+}
+
+#[derive(Debug)]
+pub struct RawWriteDecoder<'a> {
+    header: RaftRequestHeader,
+    buf: &'a [u8],
+}
+
+impl<'a> RawWriteDecoder<'a> {
+    pub fn new(buf: &'a [u8]) -> Result<RawWriteDecoder<'a>, RaftCmdRequest> {
+        match buf.first().cloned() {
+            Some(MAGIC_PREFIX) => {
+                let mut is = CodedInputStream::from_bytes(&buf[1..]);
+                let header = is.read_message().unwrap();
+                let read = is.pos();
+                Ok(RawWriteDecoder {
+                    header,
+                    buf: &buf[1 + read as usize..],
+                })
+            }
+            _ => {
+                let mut req = RaftCmdRequest::new();
+                req.merge_from_bytes(buf).unwrap();
+                Err(req)
+            }
+        }
+    }
+
+    #[inline]
+    pub fn header(&self) -> &RaftRequestHeader {
+        &self.header
+    }
+}
+
+impl<'a> Iterator for RawWriteDecoder<'a> {
+    type Item = RawWrite<'a>;
+
+    #[inline]
+    fn next(&mut self) -> Option<Self::Item> {
+        decode(&mut self.buf)
+    }
+}
+
+const PUT_TAG: u8 = 0;
+const DELETE_TAG: u8 = 1;
+const DELETE_RANGE_TAG: u8 = 2;
+
+const DEFAULT_CF_TAG: u8 = 0;
+const WRITE_CF_TAG: u8 = 1;
+const LOCK_CF_TAG: u8 = 2;
+const ARBITRARY_CF_TAG: u8 = 3;
+
+// Generally the length of most key is within 128. The length of value is
+// within 2GiB.
+#[inline]
+fn encode_len(len: u32, buf: &mut Vec<u8>) {
```

Comment:
Add comment to clarify the encoding is variable length integer encoding?
Or maybe can just use https://docs.rs/vu64/latest/vu64/

Reply:
I change the codec to sqlite4's version, which encodes small numbers more efficiently. I also add a case to test all the boundaries. It's only tens line of code, better not to introduce a new dependency.